File: socket/basic_tcp_echo/socketserver.cpp
#include <iostream>
#include <cstring>
#include <unistd.h>
#include <arpa/inet.h>

#define PORT 65456
#define BUFFER_SIZE 1024

void MyTCPSocketHandler(int clientSocket, sockaddr_in clientAddr) {
    char buffer[BUFFER_SIZE];

    std::cout << "> client connected by IP address "
              << inet_ntoa(clientAddr.sin_addr)
              << " with Port number "
              << ntohs(clientAddr.sin_port) 
              << std::endl;

    while(true) {
        memset(buffer, 0, BUFFER_SIZE);
        int bytesReceived = recv(clientSocket, buffer, BUFFER_SIZE, 0);
        if(bytesReceived <= 0) break;

        std::cout << "> echoed: " << buffer << std::endl;
        send(clientSocket, buffer, BUFFER_SIZE, 0);
        if(strcmp(buffer, "quit") == 0) break;
    }

    close(clientSocket);
    std::cout << "> client is de-activated" << std::endl;
}

int main() {
    int serverSocket, clientSocket;
    sockaddr_in serverAddr, clientAddr;
    socklen_t clientAddrSize = sizeof(clientAddr);

    serverSocket = socket(AF_INET, SOCK_STREAM, 0);
    if(serverSocket == -1) {
        std::cerr << "Error: Socket creation failed" << std::endl;
        close(serverSocket);
        return 1;
    }

    serverAddr.sin_family = AF_INET;
    serverAddr.sin_addr.s_addr = INADDR_ANY;
    serverAddr.sin_port = htons(PORT);

    if(bind(serverSocket, (sockaddr*) &serverAddr, sizeof(serverAddr)) == -1) {
        std::cerr << "Error: bind failed" << std::endl;
        close(serverSocket);
        return 1;
    }

    if(listen(serverSocket, 5) == -1) {
        std::cerr << "Error: listen failed" << std::endl;
        close(serverSocket);
        return 1;
    }

    std::cout << "> echo-server is activated" << std::endl;

    clientSocket = accept(serverSocket, (sockaddr*) &clientAddr, &clientAddrSize);
    if(clientSocket == -1) {
        std::cerr << "Error: accpet failed" << std::endl;
        close(serverSocket);
        return 1;
    }

    MyTCPSocketHandler(clientSocket, clientAddr);

    close(serverSocket);

    std::cout << "> Server is de-activated" << std::endl;
    return 0;
}

File: socket/basic_tcp_echo/socketserver.py
import socketserver

class MyTCPSocketHandler(socketserver.BaseRequestHandler):
   
    def handle(self):
        print('> client connected by IP address {0} with Port number {1}'.format(self.client_address[0], self.client_address[1])) # client_address는 튜플
        while True:
            # [=start=]
            RecvData = self.request.recv(1024)
            print('> echoed:', RecvData.decode('utf-8'))
            self.request.sendall(RecvData)
            if RecvData.decode('utf-8') == 'quit':
                break
            # [==end==]

if __name__ == "__main__":
    HOST, PORT = '127.0.0.1', 65456
    print('> echo-server is activated')
    # Create the server, binding to localhost on port 9999
    with socketserver.TCPServer((HOST, PORT), MyTCPSocketHandler) as server:
        server.serve_forever()
    print('> echo-server is de-activated')
